KAFKA-4548: Add CompatibilityTest to verify that individual features …

cmccabe opened the pull request:
Add CompatibilityTest to verify that individual features are supported or not by the broker we're connecting to

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -456,6 +456,20 @@ public static String formatAddress(String host, Integer port) {
         return sb.toString();
     }
 
+    public static <K, V> String join(Map<K, V> map,
```
There is a `mkString` method in the same class so maybe we should use that name to be consistent. Or we could rename the other one to `join` (which is what the Java standard library uses).

cmccabe replied:
Hmm, interesting.  It looks like Utils.java already has both Utils#mkString(Collection<T> coll, String separator) and Utils#join(Collection<T> list, String seperator).  The mkString version returns a CharSequence instead of a String.  I'm guessing that the reason is increased efficiency-- the StringBuilder object itself implements CharSequence, whereas converting a StringBuilder into a String requires copying.  This could get a little tricky, so perhaps we should leave it alone for now, or at least open a separate JIRA for it?

cmccabe commented:
Fix checkstyle

cmccabe committed:
KAFKA-4548. Add CompatibilityTest to verify that individual features are supported or not by the broker we're connecting to

cmccabe commented:
Rebase on trunk

cmccabe commented:
I fixed some checkstyle warnings.

I'm not sure happened with Jenkins... I saw ":core:test FAILED" and "Process 'Gradle Test Executor 3' finished with non-zero exit value 1" but there were no failed tests.  :core:test should be unaffected by this patch, which doesn't change the core directory anyway...

cmccabe commented:
Fix checkstyle

ewencp commented:
@cmccabe Seems this patch will need a rebase.

ijuma commented on tests/kafkatest/tests/client/run_compatibility_test.py:
```diff
@@ -0,0 +1,92 @@
+
+def get_broker_features(broker_version):
+    features = {}
+    if (broker_version < V_0_10_0_1):
```
I think this should be `V_0_10_1_0`.

cmccabe replied:
good catch

ijuma commented on tests/kafkatest/tests/client/run_compatibility_test.py:
```diff
@@ -0,0 +1,92 @@
+          self.logger.info("** Command failed.  See %s for log messages." % ssh_log_file)
+          raise e
+
+    # TODO: when KAFKA-4462 is complete, we should test other supported versions here.
```
KAFKA-4462 is complete. :) Maybe we can take the chance to update `TestProducerConsumerCompat` as well.

cmccabe replied:
Yeah, we should be testing older versions here now.  I'll update test_producer_consumer_compat.py as well.

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,256 @@
+                System.exit(1);
+            }
+        }
+        String bootstrapServers = res.getString("bootstrap_servers");
```
We are trying to make all tools consistent with regards to this and the name that most use is `bootstrap-server`.

cmccabe replied:
k

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,256 @@
+        }
+        String bootstrapServers = res.getString("bootstrap_servers");
+        String topic = res.getString("topic");
+        boolean offsetsForTimesSupported = res.getBoolean("offsetsForTimesSupported");
```
We typically use dash separated names for command-line parameters.

cmccabe replied:
ok

ijuma reviewed (commented):
Thanks for the PR, left a few comments.

cmccabe committed:
address review comments

run_compatibility_test: fix issue where we were checking for
broker_version < V_0_10_0_1 instead of V_0_10_1_0 when determining
whether the offsetsForTimes and clusterId features were supported.

run_compatibility_test, test_producer_consumer_compat: test older
versions

CompatibilityTest: use dash-separated arguments instead of camelcase

cmccabe commented:
Rebase on trunk

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -93,22 +93,22 @@ public static void main(String[] args) throws Exception {
             }
         }
-        String bootstrapServers = res.getString("bootstrap_servers");
+        String bootstrapServer = res.getString("bootstrap_server");
```
This should be `bootstrap-server`.

cmccabe replied:
It's just the name of the metavar.  The actual command line option is still --bootstrap-server.  Confusing, I know, but dashes in argparse4j metavars don't seem to work (curious if anyone has any more information on this limitation)

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -93,22 +93,22 @@ public static void main(String[] args) throws Exception {
+        String bootstrapServer = res.getString("bootstrap_server");
         String topic = res.getString("topic");
-        boolean offsetsForTimesSupported = res.getBoolean("offsetsForTimesSupported");
+        boolean offsetsForTimesSupported = res.getBoolean("offsets_for_times_supported");
```
Hmm, should be dash-separated, right? Also for the cluster id a few lines below.

cmccabe replied:
same here

ijuma commented on tests/kafkatest/tests/client/test_producer_consumer_compat.py:
```diff
@@ -53,8 +53,10 @@ def min_cluster_size(self):
 
-    # TODO: when KAFKA-4462 is fully implemented, we should test other versions here.
     @parametrize(broker_version=str(TRUNK))
+    @parametrize(broker_version=str(V_0_10_0_0))
```
We tend to test the latest version in a branch, say `LATEST_0_10_0` and `LATEST_0_10_1`. Do you think that would be enough? Otherwise the number of variants will increase quite rapidly.

cmccabe replied:
OK, let's just test those.  We shouldn't add major features in point releases anyway.

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    private static class CompatibilityTestSenderCompletion implements Callback {
+        private CountDownLatch latch = new CountDownLatch(1);
```
Should be final?

cmccabe replied:
yes

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,256 @@
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
+                "org.apache.kafka.common.serialization.ByteArraySerializer");
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+                "org.apache.kafka.common.serialization.ByteArraySerializer");
```
Instead of setting it this way, you can just pass the serializers to the constructor.

cmccabe replied:
good idea

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,256 @@
+        String topic = res.getString("topic");
+        boolean offsetsForTimesSupported = res.getBoolean("offsets_for_times_supported");
+        CompatibilityTestDeserializer.expectClusterId =
+                res.getBoolean("cluster_id_supported");
```
We seem to be using underscores instead of dashes in a few places still.

cmccabe replied:
ditto

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,256 @@
+        consumerProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+                "org.apache.kafka.tools.CompatibilityTest$CompatibilityTestDeserializer");
+        consumerProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+                "org.apache.kafka.common.serialization.ByteArrayDeserializer");
```
You can pass the deserializer in the constructor. That way you can also avoid the static variable in favour of a constructor parameter.

cmccabe replied:
good idea

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,256 @@
+        CompatibilityTestSenderCompletion cb =
+                new CompatibilityTestSenderCompletion();
+        KafkaProducer<byte[], byte[]> producer = new KafkaProducer<byte[], byte[]>(props);
+        producer.send(record, cb);
```
Why not call `get` on the future instead of using the callback handler?

cmccabe replied:
ok

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,256 @@
+    public static void testConsume(String topic, Properties props,
+                                   boolean offsetsForTimesSupported) throws Exception {
+        byte[] payload = new byte[] {0x61, 0x62, 0x63};
+        CompatibilityTestSenderCompletion cb =
```
Seems unused.

cmccabe replied:
removed

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,256 @@
+                new Runnable() {
+                    @Override
+                    public void run() {
+                        System.out.println("offsetsForTime = " + offsetsForTime.result);
```
I didn't quite understand the point of this. Is it just a debugging aid?

cmccabe replied:
it is just for debugging.

ijuma reviewed (commented):
Thanks for the updates, took another pass.

cmccabe committed:
refactors, test LATEST_0_10_0 and LATEST_0_10_1 in ducktape

cmccabe commented:
I did some refactoring to make it nicer.  I also added the consumer reading back the messages we wrote.

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,383 @@
+            t.printStackTrace();
+            System.exit(1);
+        }
+        System.out.printf("SUCCESS.\n");
```
Is there a reason why we are using `printf` instead of `println` here?

cmccabe replied:
changed to println

cmccabe committed:
CompatibilityTestSenderCompletion: use Future#get instead of callback to register completion

cmccabe committed:
fix checkstyle

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,359 @@
+import java.util.Properties;
+import java.util.concurrent.Future;
+
+public class CompatibilityTest {
```
Should this be called `ClientCompatibilityTest`? Also, probably worth adding a brief comment about this being used by system tests and not meant to be used as a tool. We could also perhaps move it to a subpackage to make that clear.

cmccabe replied:
ClientCompatibilityTest seems like a good name.  I will add some JavaDoc to the test as well.

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,359 @@
+            timestampsToSearch.put(topicPartition, prodTimeMs);
+            topicPartitions.add(topicPartition);
+        }
+        final OffsetsForTime offsetsForTime = new OffsetsForTime();
```
It seems like we don't use this anywhere?

cmccabe replied:
It's used in the tryFeature(...) block

ijuma replied:
Yes, I know we assign `result` in the block, but I don't understand what the aim is as we don't do anything else with it. It seems like we can remove the assignment and the behaviour remains the same. What am I missing?

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,359 @@
+
+        consumer.assign(topicPartitions);
+        consumer.seekToBeginning(topicPartitions);
+        final Iterator<byte[]> iter = new Iterator<byte[]>() {
```
Why do we need this? Can we not simply do a while loop in `fetchNext` and accumulate the results into a collection?

cmccabe replied:
There is a follow-up change which makes the second message larger than the bytes per partition maximum, in order to test the KIP-74 behavior.  We're not doing the same thing with every message

ijuma replied:
OK, we can leave as is then for now.

ijuma commented on tools/src/main/java/org/apache/kafka/tools/CompatibilityTest.java:
```diff
@@ -0,0 +1,359 @@
+            compareArrays(message2, next);
+            log.debug("Found second message...");
+        } catch (RuntimeException e) {
+            throw new RuntimeException("The first message in this topic was not ours.  Please use a new " +
```
I think you meant `second` message here. Why not simply accumulate the results into a collection of Strings and then do `arrayEquals(List("foo1", "foo2"), messages)` or something like that? It seems like the code would be simpler.

cmccabe replied:
See above

ijuma reviewed (commented):
Thanks for the updates, left a few comments.

cmccabe committed:
Rename CompatibilityTest -> ClientCompatibilityTest, add JavaDoc

cmccabe committed:
printf -> println, fix typo

ijuma commented:
I ran this in Jenkins and there was one failure that looks related:

```
test_id:    kafkatest.tests.client.test_producer_consumer_compat.TestProducerConsumerCompat.test_produce_consume.broker_version=0.10.1.0
status:     FAIL
run time:   54.753 seconds


    Kafka server didn't finish startup
```
https://jenkins.confluent.io/job/system-test-kafka-branch-builder/675/console

I started another build in case it was an environmental issue:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/678/

ijuma commented on tests/kafkatest/tests/client/run_client_compatibility_test.py:
```diff
@@ -45,15 +45,15 @@ def run_command(node, cmd, ssh_log_file):
 
 
-class RunCompatibilityTest(Test):
+class RunClientCompatibilityTest(Test):
```
Nit: I think I'd just call this `ClientCompatibilityTest`. The fact that it runs the Java tool to do that is a bit of an implementation detail, isn't it?

ijuma commented:
Seems like the same system test failed again.

cmccabe committed:
run_client_compatibility_test.py: fix indentation problem

cmccabe commented:
The issue with 0.10.1.0 failing to start up appears to be https://issues.apache.org/jira/browse/KAFKA-4688 .  Let's just remove that version from the tested list until KAFKA-4688 can be committed and our environments are updated.

cmccabe commented:
I located the failing log message here from Kafka:

http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2017-01-21--001.1485018206--cmccabe--KAFKA-4548--6d2094c/TestProducerConsumerCompat/test_produce_consume/broker_version%3D0.10.1.0/4.tgz

It is a single line:
```
bash: /opt/kafka-0.10.1.0/bin/kafka-server-start.sh: No such file or directory
```

So this confirms that the problem is the test environment, not the test.

cmccabe commented:
I removed 0.10.1.0 from the list of tested versions.  Once KAFKA-4688 is in, we can re-add it.  I also rebased on trunk.  Thanks for the reviews, all.

cmccabe committed:
Fix run_client_compatibility_test.py to use a custom topic name, and use os.mkdirs on log dir for ssh error file

cmccabe committed:
run_client_compatibility_test: use only one partition to avoid worrying about ordering

ijuma committed:
Renames and minor clean-ups

cmccabe commented:
I re-added 0.10.1.0 as a tested version since KAFKA-4688 got committed.  I ran the tests inside docker and they all passed:

```
+ docker exec knode01 bash -c 'cd /kafka_src; ducktape  --cluster-file tests/cluster_file.json tests/kafkatest/tests/client/run_client_compatibility_test.py'
[INFO:2017-01-24 17:46:58,625]: starting test run with session id 2017-01-24--001...
[INFO:2017-01-24 17:46:58,625]: running 3 tests...
[INFO:2017-01-24 17:46:58,625]: Triggering test 1 of 3...
[INFO:2017-01-24 17:46:58,839]: RunnerClient: Loading test {'directory': '/kafka_src/tests/kafkatest/tests/client', 'file_name': 'run_client_compatibility_test.py', 'method_name': 'run_compatibility_test', 'cls_name': 'RunClientCompatibilityTest', 'injected_args': {'broker_version': '0.10.0.1'}}
[INFO:2017-01-24 17:46:58,852]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.0.1: Setting up...
[INFO:2017-01-24 17:46:58,853]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.0.1: Running...
j[INFO:2017-01-24 17:47:36,621]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.0.1: PASS
[INFO:2017-01-24 17:47:36,622]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.0.1: Tearing down...
[INFO:2017-01-24 17:48:01,347]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.0.1: Summary: 
[INFO:2017-01-24 17:48:01,348]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.0.1: Data: None
[INFO:2017-01-24 17:48:01,356]: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[INFO:2017-01-24 17:48:01,356]: Triggering test 2 of 3...
[INFO:2017-01-24 17:48:01,366]: RunnerClient: Loading test {'directory': '/kafka_src/tests/kafkatest/tests/client', 'file_name': 'run_client_compatibility_test.py', 'method_name': 'run_compatibility_test', 'cls_name': 'RunClientCompatibilityTest', 'injected_args': {'broker_version': '0.10.1.1'}}
[INFO:2017-01-24 17:48:01,372]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.1.1: Setting up...
[INFO:2017-01-24 17:48:01,373]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.1.1: Running...
[INFO:2017-01-24 17:48:31,580]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.1.1: PASS
[INFO:2017-01-24 17:48:31,581]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.1.1: Tearing down...
[INFO:2017-01-24 17:48:58,955]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.1.1: Summary: 
[INFO:2017-01-24 17:48:58,956]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=0.10.1.1: Data: None
[INFO:2017-01-24 17:48:58,966]: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[INFO:2017-01-24 17:48:58,967]: Triggering test 3 of 3...
[INFO:2017-01-24 17:48:58,978]: RunnerClient: Loading test {'directory': '/kafka_src/tests/kafkatest/tests/client', 'file_name': 'run_client_compatibility_test.py', 'method_name': 'run_compatibility_test', 'cls_name': 'RunClientCompatibilityTest', 'injected_args': {'broker_version': 'trunk'}}
[INFO:2017-01-24 17:48:58,982]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=trunk: Setting up...
[INFO:2017-01-24 17:48:58,984]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=trunk: Running...
[INFO:2017-01-24 17:49:30,096]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=trunk: PASS
[INFO:2017-01-24 17:49:30,097]: RunnerClient: kafkatest.tests.client.run_client_compatibility_test.RunClientCompatibilityTest.run_compatibility_test.broker_version=trunk: Tearing down...
```

cmccabe commented:
I ran the ducktape tests in Docker and they all passed:

```
KAFKA_IMAGE_REBUILD=1 TC_PATHS="tests/kafkatest/tests/client/client_compatibility_features_test.py" bash ./tests/docker/run_tests.sh
KAFKA_IMAGE_REBUILD=1 TC_PATHS="tests/kafkatest/tests/client/client_compatibility_produce_consume_test.py" bash ./tests/docker/run_tests.sh
```
